StoredProcedureEFCore.Tests.SqlServer/Program.cs

```diff
@@ -87,16 +90,160 @@ await ctx.LoadStoredProc("dbo.OutputNullable")
+            await TestCancelAfterResultset(ctx, 2, async (r, ct) => await r.ToListAsync<Model>(ct));
+
+            await TestCancelBeforeResultset(ctx, 3, async (r, ct) => await r.ColumnAsync<Int64>("id", ct));
+            await TestCancelAfterResultset(ctx, 4, async (r, ct) => await r.ColumnAsync<Int64>("id", ct));
+
+            await TestCancelBeforeResultset(ctx, 5, async (r, ct) => await r.ToDictionaryAsync<Int64, Model>(m => m.Id, ct));
+            await TestCancelAfterResultset(ctx, 6, async (r, ct) => await r.ToDictionaryAsync<Int64, Model>(m => m.Id, ct));
+
+            await TestCancelBeforeResultset(ctx, 7, async (r, ct) => await r.ToLookupAsync<Int64, Model>(m => m.Id, ct));
+            await TestCancelAfterResultset(ctx, 8, async (r, ct) => await r.ToLookupAsync<Int64, Model>(m => m.Id, ct));
+
+            await TestCancelBeforeResultset(ctx, 9, async (r, ct) => await r.ToSetAsync<Int64>(ct));
+            await TestCancelAfterResultset(ctx, 10, async (r, ct) => await r.ToSetAsync<Int64>(ct));
+
+            await TestCancelBeforeResultset(ctx, 11, async (r, ct) => await r.FirstAsync<Model>(ct));
+            await TestCancelAfterResultset(ctx, 12, async (r, ct) => await r.FirstAsync<Model>(ct));
+
+            await TestCancelBeforeResultset(ctx, 13, async (r, ct) => await r.SingleAsync<Model>(ct), 1);
+            await TestCancelAfterResultset(ctx, 14, async (r, ct) => await r.SingleAsync<Model>(ct), 1);
+
+            await TestCancelBeforeResultset(ctx, 15, async (r, ct) => await r.SingleOrDefaultAsync<Model>(ct), 1);
+            await TestCancelAfterResultset(ctx, 16, async (r, ct) => await r.SingleOrDefaultAsync<Model>(ct), 1);
+        }
+
+#pragma warning disable AsyncFixer04 // A disposable object used in a fire & forget async call
+        /// <summary>
+        /// Test async cancellation where the sproc is cancelled before it has a chance to produce a resultset.
+        /// </summary>
+        static async Task TestCancelBeforeResultset(DbContext ctx, int callId, Func<DbDataReader, CancellationToken, Task> action, Int64 limit = 9223372036854775807)
+        {
+            var stopWatch = new Stopwatch();
+            stopWatch.Start();
+
+            try
+            {
+                Console.WriteLine("\n====================================");
+                Console.WriteLine($"Starting long-running call #{callId}...");
+
+                using (var cts = new CancellationTokenSource())
+                {
+                    var sprocTask = ctx.LoadStoredProc("dbo.ListAll")
+                        .AddParam("limit", limit)
+                        .AddParam("delay_in_seconds_before_resultset", 10)
+                        .ExecAsync(async r => await action(r, cts.Token), cts.Token);
+
+                    // Cancel the task after one second.
+                    var cancelTask = CancelTokenAsync(callId, 1000, cts);

```

Comment:
Since the task is never awaited, can't you use [CancelAfter](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.cancelafter)?

Reply:
I don't know what you mean by, "the task is not awaited".
Please see next line `await Task.WhenAll(...)`.